text/3748-cargo_check_environment_variable.md

```diff
@@ -0,0 +1,107 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new environment variable `CARGO_CHECK` that is set to `1` when running `cargo check` or similar type-checking operations so build scripts can skip expensive compilation steps that are unnecessary for Rust type checking, such as compiling external C++ code in cxx based projects.
+
+# Motivation
+[motivation]: #motivation
+
+Rust development heavily relies on IDE tooling like rust-analyzer, which frequently invokes `cargo check` to provide real-time type information and diagnostics. Many projects use build scripts (`build.rs`) to generate Rust code and compile external dependencies. For example:
+
+- cxx-rs generates Rust bindings for C++ code and compiles C++ source files
+- cxx-qt generates Rust bindings for Qt code and runs the Qt Meta-Object Compiler (MOC)
+- Projects using Protocol Buffers generate Rust code from .proto files
+- bindgen generates Rust bindings from C/C++ headers
+
+Currently, every time rust-analyzer runs `cargo check`, the build script in the changed crate must execute its full build process, including steps like compiling C++ code that are only needed for linking but not for type checking. Normally the build script would only be run when a file added by `cargo::rerun-if-changed` is changed, which generally doesn't include the Rust source code. However, when using `cxx` to create bridges between C++ and Rust, the build script must be run for every change in the Rust bridges. Usually `cxx` bindings are rarely changed but in projects like `cxx-qt` that interface between Rust and Qt types, they receive signficantly more changes. This impacts IDE responsiveness, especially in projects with complex build scripts.
+
+This is particularly important for projects using cxx-qt and similar frameworks where the build scripts perform extensive code generation and compilation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing a build script (`build.rs`), you can now check the `CARGO_CHECK` environment variable to determine if the build is being performed for type checking purposes:
+
+```rust
+fn main() {
+    generate_rust_bindings();
+
+    // Only compile external code when not type checking
+    if std::env::var("CARGO_CHECK").is_ok() {
+        compile_cpp_code();
+    }
+}
+```
+
+This allows build scripts to optimize their behavior based on the build context. When rust-analyzer or a developer runs `cargo check`, the build script can skip time-consuming steps that aren't necessary for type checking.
+
+This feature primarily benefits library authors who maintain build scripts, especially those working with external code generation and compilation. Regular Rust developers using these libraries will automatically benefit from improved IDE performance without needing to modify their code.
+
+# Reference-level explanation
```

Comment:
This should address things like caching and output generation. Quoting @alexcrichton in https://github.com/rust-lang/cargo/issues/4001#issuecomment-983000999:

> I'm not 100% certain this still fits in Cargo myself. The major downside of implementing a feature like this is that cargo check && cargo build gets slower than it currently is. Cargo currently caches build script invocations for those two commands, which means that all the work done by cargo check is reused by cargo build and isn't redone.
>
> [...]
>
> As an author of crates like cc and cmake it's also somewhat ambiguous to me about what "check mode" would do for libraries like that. Should they do nothing? Type-check the C code? Make sure it all compiles? Basically I don't think that C/C++ have any real meaningful distinction like Rust does for cargo check and cargo build, so there's not really an obvious choice of what these library crates would do, which would require even more opt-in or configuration on behalf of all users.

High level proposal: for a build script, `CHECK_ONLY` should mean:

1. Do any relevant checks that don't include codegen, if possible
2. Do what is necessary to check the _Rust_ code
3. This can always be disregarded in favor of running the full build script

So for C code, `CHECK_ONLY` would mean doing the C/++ configure (but not build) step and possibly invoking bindgen. Usually this is `./configure` or `cmake` without `--build` to verify dependencies and environment, and prepare final headers and source files. `bindgen` can then be run on these final headers. `!CHECK_ONLY` then means that `make`, `cmake --build`, or `cc` should be run.

`OUT_DIR` should be the same between check and build so configuration output can be reused. The build script will probably always run the congigure step, `cmake` knows how to skip a lot if files are already up to date [^1].

For crates, `cc` should totally ignore this environment variable and not change anything - the build script author will just be have the option to skip invoking `cc` if `CHECK_ONLY` is set. `cmake`'s `Config` should add new methods like `only_configure` (always called) and `only_build` (called if `CHECK_ONLY` is unset), since `Config::build` currently does both these things. `cmake::build` and `Config::build` could optionally listen to `CHECK_ONLY` to determine whether to run the `only_build` step.

[^1]: Cmake's configure caching can sometimes be pretty slow for bigger projects. It would be possible for build scripts to thumbprint relevant environment and skip invoking `cmake` if `OUT_DIR` has an updated thumbprint, but this would be advanced usage.